### PR TITLE
fix stable diffusion fine-tune backward graph capture and lowering

### PR DESCRIPTION
**Pull request by Yancey1989**

### CODE REVIEW

File: pytorch_blade/pytorch_blade/compiler/jit/torch/shape_analysis.cpp [hunk 1]
```diff
@@ -2700,6 +2861,266 @@ class ShapePropagator : public PropertyPropBase {
     }
     return false;
   }
+#if PYTORCH_VERSION_GE(1, 14)
+  bool PropagateCompleteShapeOnNode(
+      Node* node,
+      bool insert_expands,
+      std::vector<TensorTypePtr> tensor_types) {
+    node->dump();
```
Comment (tanyokwok) [inline pytorch_blade/pytorch_blade/compiler/jit/torch/shape_analysis.cpp:2869]: Nit: remove the dump?

File: pytorch_blade/pytorch_blade/torch-mlir/lib/Conversion/TorchToMhlo/DiscTorchToMhlo.cpp [hunk 2]
```diff
@@ -671,6 +671,55 @@ LogicalResult ConvertAtenOp<AtenNegIntOp>::matchAndRewrite(
...
+    return op.emitError("only Tensor types supported in MHLO");
+  auto loc = op.getLoc();
+  Type outElemTy = outType.getElementType();
+  if (!outElemTy.isIntOrFloat())
+    return op.emitError(
+        "only floating-point or integer datatype legalization supported");
+  auto atenEMFOp =
+      dyn_cast<AtenEmptyMemoryFormatOp>(op.getSelf().getDefiningOp());
+  if (!atenEMFOp) {
+    return op.emitError("the producer of fill.Scaler should be aten.empty.");
```
Comment (tanyokwok) [inline pytorch_blade/pytorch_blade/torch-mlir/lib/Conversion/TorchToMhlo/DiscTorchToMhlo.cpp:692]: Typo: fill.Scalar

File: pytorch_blade/pytorch_blade/compiler/jit/torch/shape_analysis.cpp [hunk 3]
```diff
@@ -2700,6 +2861,266 @@ class ShapePropagator : public PropertyPropBase {
     }
     return false;
   }
+#if PYTORCH_VERSION_GE(1, 14)
+  bool PropagateCompleteShapeOnNode(
```
Comment (tanyokwok) [inline pytorch_blade/pytorch_blade/compiler/jit/torch/shape_analysis.cpp:2865]: When PropagateCompleteShapeOnNode should be used?

**Comment by tanyokwok** [comment]
The PR has done a bunch of things. Can you split it into several more small and concise ones?
1. Add more shape analysis functions and file checks
2. Add PropagateCompleteShapeOnNode for some purpose
3. Add const folding for Torch2.0 training
4. Add custom ops env variables
5. And so on

**Comment by Yancey1989** [comment]
Thanks @tanyokwok , after the offline discussion, I repeated this PR into 3 smaller ones, #1111 #1112 and #1113 , please take a look again.

**Pull request closed without merging**